src/Functions/if.cpp

```diff
@@ -1186,24 +1197,48 @@ class FunctionIf : public FunctionIfBase
     ColumnNumbers getArgumentsThatDontImplyNullableReturnType(size_t /*number_of_arguments*/) const override { return {0}; }
     bool canBeExecutedOnLowCardinalityDictionary() const override { return false; }
 
-    /// Get result types by argument types. If the function does not apply to these arguments, throw an exception.
-    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+
+    static void checkConditionArgType(const DataTypePtr & type)
     {
-        if (!arguments[0]->onlyNull())
+        if (!type->onlyNull())
         {
-            if (arguments[0]->isNullable())
-                return getReturnTypeImpl({
-                    removeNullable(arguments[0]), arguments[1], arguments[2]});
+            if (type->isNullable())
+            {
+                checkConditionArgType(removeNullable(type));
+                return;
+            }
+            if (!WhichDataType(type).isUInt8())
+                throw Exception(
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                    "Illegal type {} of first argument (condition) of function if. "
+                    "Must be UInt8.",
+                    type->getName());
+        }
+    }
 
-            if (!WhichDataType(arguments[0]).isUInt8())
-                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of first argument (condition) of function if. "
-                    "Must be UInt8.", arguments[0]->getName());
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        checkConditionArgType(arguments[0].type);
+
+        // Some processing - for constant strings - for LowCardinality
+        if (use_low_cardinality_optimisation && arguments[1].column && arguments[2].column && isColumnConst(*arguments[1].column)
+            && isColumnConst(*arguments[2].column))
+        {
+            auto const is_string1 = isString(arguments[1].type);
+            auto const is_string2 = isString(arguments[2].type);
+            if (is_string1 && is_string2)
+                return std::make_shared<DataTypeLowCardinality>(std::make_shared<DataTypeString>());
+            // Still, might be a mix of string + null (or vice-versa)
+            auto const is_null1 = arguments[1].type->onlyNull();
+            auto const is_null2 = arguments[2].type->onlyNull();
+            if ((is_string1 && is_null2) || (is_string2 && is_null1))
+                return std::make_shared<DataTypeLowCardinality>(std::make_shared<DataTypeNullable>(std::make_shared<DataTypeString>()));
```

Comment:
I was wondering if there's anything else to be done apart from figuring out the case for `LowCardinality` in `getReturnTypeImpl`